File: books/halim/src_uva/uva_00787.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <set>

using namespace std;

// Constants
// Simple types
#define ll long long
#define ld long double

// Aggeragte types
#define pii pair<int,int>
#define pll pair<long,long>

#define vi vector<int>
#define vl vector<long>

#define si set<int>
#define sl set<long>

// Output
template <typename T>
void print_v(vector<T>& v) {cout << "{"; for (auto& x : v) cout << x << " "; cout << "\n";}

int maxProd(vector<int>& nums) {
    int max_ = nums[0];
    int cNeg = nums[0];
    int cPos = nums[0];
    for (int i = 1; i < nums.size(); ++i) {
        int oPos = cPos, oNeg = cNeg;
        cPos = max(max(nums[i], cPos * nums[i]), oNeg * nums[i]);
        cNeg = min(min(nums[i], cNeg * nums[i]), oPos * nums[i]);       
        max_ = max(max_, cPos);
    }
    return max_;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
    vector<int> numbers;
    int num;
    while (scanf("%d", &num) != EOF) {
        if (num == -999999) {
            cout << maxProd(numbers) << "\n";
            numbers.clear();
            continue;
        }
        numbers.push_back(num);
    }    

	return 0;
}

File: books/halim/src_uva/UVa_00787.java
import java.util.LinkedList;
import java.util.Scanner; 
import java.util.*;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.math.BigInteger;

class Main {

    public static BigInteger result(BigInteger list[], int l) {
        BigInteger max_ = list[0];
        BigInteger cNeg = list[0];
        BigInteger cPos = list[0];

        for (int i = 1; i < l; ++i) {
            BigInteger oPos = cPos;
            BigInteger oNeg = cNeg;

            cPos = oNeg.multiply(list[i])
                    .max(
                        list[i].max(
                            cPos.multiply(list[i])
                        )
                    );
            cNeg = oPos.multiply(list[i])
            .min(
                list[i].min(
                    cNeg.multiply(list[i])
                )
            ); 
            max_ = max_.max(cPos);
        }

        return max_;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        BigInteger[] array = new BigInteger[105];
        int l = 0;
        while (scanner.hasNextInt()) {
            int i = scanner.nextInt();
            if (i == -999999) {
                BigInteger r = result(array, l);
                System.out.println(r);
                l = 0;
            } else {
                array[l++] = new BigInteger(String.valueOf(i));
            }
        }
    }
}
